q2_annotate/busco/busco.py

```diff
@@ -30,7 +30,139 @@
+    """
+    Partition unbinned contigs to match the partitioning of MAGs.
+    
+    Args:
+        unbinned (ContigSequencesDirFmt): The unbinned contigs directory.
+        mags_partitioned (MultiMAGSequencesDirFmt): The partitioned MAGs structure.
+        
+    Returns:
+        Dict[str, ContigSequencesDirFmt]: Mapping of partitions to unbinned contigs.
+    """
+    # unbinned = Artifact.load(unbinned_qza).view(ContigSequencesDirFmt)
+    partitioned_unbinned = {}
+
+    for partition_id, mag_partition in mags_partitioned.items():  # Loop over MAG partitions
+        result = ContigSequencesDirFmt()  # Temporary storage for partitioned unbinned contigs
+        for sample_id in mag_partition.sample_dict().keys():  # Extract sample names from MAGs
+            expected_unbinned_filename = f"{sample_id}_contigs.fa"
+            unbinned_sample_dict = unbinned.sample_dict()  # Get unbinned sample mapping
+            
+            if sample_id in unbinned_sample_dict:  # Check if unbinned contigs exist for this sample
+                unbinned_fp = unbinned_sample_dict[sample_id]
+                os.makedirs(result.path, exist_ok=True)
+                duplicate(unbinned_fp, result.path / expected_unbinned_filename)
+        
+        partitioned_unbinned[partition_id] = result  # Store the partitioned unbinned contigs
+    print("\nFinal Partitioned unbinned Structure:")
+    print(partitioned_unbinned)
+    return partitioned_unbinned
+def count_binned_contigs(bins) -> int:
+    """Counts sequences in all FASTA files inside a MultiFASTADirectoryFormat using DNAIterator."""
+    
+    total_sequences = 0
+
+    #Iterate over sequences directly using QIIME 2's DNAIterator
+    for fasta_fp, dna_iterator in bins.sequences.iter_views(DNAIterator):
+        sequence_count = sum(1 for _ in dna_iterator)  # Count sequences
+        # print(f"{fasta_fp} contains {sequence_count} sequences")  # Debugging output
+        total_sequences += sequence_count
+
+    return total_sequences
+
+def count_unbinned_contigs(unbinned_contigs) -> int:
+    """Counts sequences in the unbinned FASTA file using DNAIterator."""
+    
+    total_sequences = 0
+
+    for fasta_fp, dna_iterator in unbinned_contigs.sequences.iter_views(DNAIterator):
+        sequence_count = sum(1 for _ in dna_iterator)  # Count sequences
+        # print(f"{fasta_fp} contains {sequence_count} sequences")  # Debugging output
+        total_sequences += sequence_count
+
+    return total_sequences
+
+def calculate_unbinned_percentage(bins, unbinned_contigs):
+    """ Calculates the percentage of unbinned contigs """
+    #need to remove loading artifacts
+    
+    # # Load QIIME2 artifacts and get correct directory formats
+    # mags_dir = Artifact.load(mags_qza).view(MultiMAGSequencesDirFmt)  # Binned MAGs
+    # unbinned_dir = Artifact.load(unbinned_qza).view(ContigSequencesDirFmt)  # Unbinned Contigs
+
+    # Count sequences
+    binned_contigs = count_binned_contigs(bins)
```

Comment:
This should probably happen before this method is called, not here. If you do it here, you need to calculate this every time you process a new sample, which is unnecessary.